src/Server.php

```diff
@@ -1,95 +1,175 @@
 <?php
+/**
+ * @package yswery\DNS
+ */
 
 namespace yswery\DNS;
 
+use yswery\DNS\Event\EventSubscriberInterface;
+use yswery\DNS\Resolver\ResolverInterface;
+
+/**
+ * Class Server
+ */
 class Server
 {
+    private $ip = '0.0.0.0';
+    private $port = 53;
+    private $ttl = 300;
+    private $maxPacketLength = 512;
 
     /**
-     * @var AbstractStorageProvider
+     * @var ResolverInterface
      */
-    private $ds_storage;
+    private $resolver;
+
+    protected $eventSubscribes = [];
 
-    public function __construct($ds_storage, $bind_ip = '0.0.0.0', $bind_port = 53, $default_ttl = 300, $max_packet_len = 512)
+    /**
+     * Server constructor.
+     *
+     * @param ResolverInterface $resolver
+     * @param array             $config
+     */
+    public function __construct(ResolverInterface $resolver, array $config = [])
     {
-        $this->DS_PORT = $bind_port;
-        $this->DS_IP = $bind_ip;
-        $this->DS_TTL = $default_ttl;
-        $this->DS_MAX_LENGTH = $max_packet_len;
-        $this->ds_storage = $ds_storage;
+        // @todo validate configuration validate ip, port, max packet and ttl as positive integers
+
+        if (isset($config['server']) && $server = $config['server']) {
+            $this->ip = isset($server['bind']) ? $server['bind'] : $this->ip;
+            $this->port = isset($server['port']) ? $server['port'] : $this->port;
+            $this->ttl = isset($server['ttl']) ? $server['ttl'] : $this->ttl;
+            $this->maxPacketLength = isset($server['max_packet_length']) ? $server['max_packet_length'] : $this->maxPacketLength;
+        }
+
+        $this->resolver = $resolver;
 
         ini_set('display_errors', true);
         ini_set('error_reporting', E_ALL);
 
-        set_error_handler(array($this, 'ds_error'), E_ALL);
+        set_error_handler([$this, 'errorHandler'], E_ALL);
         set_time_limit(0);
 
         if (!extension_loaded('sockets') || !function_exists('socket_create')) {
-            $this->ds_error(E_USER_ERROR, 'Socket extension or function not found.', __FILE__, __LINE__);
+            $this->errorHandler(E_USER_ERROR, 'Socket extension or function not found.', __FILE__, __LINE__);
```

Comment:
Should this function return after this? Else it just kind of hangs and does nothing.